### PR TITLE
dbft: move payloads/block/tx/crypto interfaces to dbft package

### LINKED ISSUE #90: Consider moving dBFT payload interfaces to `dbft`
Currently generic dBFT payload interfaces are located at the `dbft/payload` folder (`ChangeView`, `PrepareRequest` and etc.). We need to consider moving them to `dbft`. Ref. https://github.com/nspcc-dev/dbft/pull/89#discussion_r1484390956

### LINKED ISSUE #91: Consider not importing default payload implementations for production applications
## Is your feature request related to a problem? Please describe.
I'm always frustrated when we import/build code we don't need. We have payloads/block implementations that are imported by default, but no real application uses them. They're useful for tests, they can be provided for edge cases, but they're not needed to be present in production builds.

## Describe the solution you'd like
Drop defaults for `New*` payload callbacks. They still can be used via `payload.*` methods, we can even add some method to provide the default set of options for all of them, but we can avoid importing them in NeoGo or Bane.

## Describe alternatives you've considered
Keep this code in our builds. Nothing critical with this, btw.

## Additional context
https://github.com/nspcc-dev/dbft/pull/89
https://github.com/nspcc-dev/dbft/issues/90

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Two parts of dBFT structure refactoring: close #90, close #91.

TODO:

- [x] Improve documentation.
- [x] Move all default implementations to testing code or separate place.
- [x] Update changelog.

### CODE REVIEW

**Comment by AnnaShaleva** [comment]
@roman-khimov, I need some initial assessment of the resulting packages structure from your side. If the direction is OK, then I'll update the documentation. The resulting package is smaller than the one got from master:
```
-rw-rw-r--  1 anna anna 189040 мар  1 22:12 build.branch
-rw-rw-r--  1 anna anna 622930 мар  1 22:11 build.master
```

I'm not sure how to deal with `internal` package. Currently `dbft` package doesn't depend on `internal` in any way, is it enough? I'm not sure if `internal` is excluded from build in this case.

File: payload/change_view.go [hunk 1]
```diff
@@ -2,29 +2,9 @@ package payload
 
 import (
 	"encoding/gob"
+	"github.com/nspcc-dev/dbft"
```
Comment (roman-khimov) [inline payload/change_view.go:6]: goimports!

File: payload/message.go [hunk 2]
```diff
@@ -3,30 +3,12 @@ package payload
 import (
 	"bytes"
 	"encoding/gob"
+	"github.com/nspcc-dev/dbft"
```
Comment (roman-khimov) [inline payload/message.go]: Can't be!

File: payload/message_test.go [hunk 3]
```diff
@@ -4,6 +4,7 @@ import (
 	"bytes"
 	"crypto/rand"
 	"encoding/gob"
+	"github.com/nspcc-dev/dbft"
```
Comment (roman-khimov) [inline payload/message_test.go]: It hurts.

File: payload/prepare_request.go [hunk 4]
```diff
@@ -2,34 +2,11 @@ package payload
 
 import (
 	"encoding/gob"
+	"github.com/nspcc-dev/dbft"
```
Comment (roman-khimov) [inline payload/prepare_request.go]: Badly.

File: payload/prepare_response.go [hunk 5]
```diff
@@ -2,19 +2,11 @@ package payload
 
 import (
 	"encoding/gob"
+	"github.com/nspcc-dev/dbft"
```
Comment (roman-khimov) [inline payload/prepare_response.go]: Bang.

File: payload/recovery_message.go [hunk 6]
```diff
@@ -3,37 +3,18 @@ package payload
 import (
 	"encoding/gob"
 	"errors"
+	"github.com/nspcc-dev/dbft"
```
Comment (roman-khimov) [inline payload/recovery_message.go]: Bang.

File: payload/recovery_request.go [hunk 7]
```diff
@@ -2,16 +2,9 @@ package payload
 
 import (
 	"encoding/gob"
+	"github.com/nspcc-dev/dbft"
```
Comment (roman-khimov) [inline payload/recovery_request.go:6]: Uh-oh.

File: simulation/main.go [hunk 8]
```diff
@@ -7,6 +7,7 @@ import (
 	"errors"
 	"flag"
 	"fmt"
+	"github.com/nspcc-dev/dbft/block"
```
Comment (roman-khimov) [inline simulation/main.go]: Meh.

File: helpers.go [hunk 9]
```diff
@@ -1,56 +1,60 @@
 package dbft
 
 type (
-	// inbox is a structure storing messages from a single epoch.
-	inbox[H Hash, A Address] struct {
-		prepare map[uint16]ConsensusPayload[H, A]
-		chViews map[uint16]ConsensusPayload[H, A]
-		commit  map[uint16]ConsensusPayload[H, A]
+	// Inbox is a structure storing messages from a single epoch.
+	Inbox[H Hash, A Address] struct {
```
Comment (roman-khimov) [inline helpers.go:5]: This I don't like. It's an internal thing. Does anything change wrt package size if you leave it as is (I know test code will have to import some implementation, but it's not compiled unless you do testing)?

File: internal/block/block_test.go [hunk 10]
```diff
@@ -6,20 +6,21 @@ import (
 	"encoding/binary"
 	"encoding/gob"
 	"errors"
-	"github.com/nspcc-dev/dbft"
 	"testing"
 
-	"github.com/nspcc-dev/dbft/crypto"
+	"github.com/nspcc-dev/dbft"
+	crypto2 "github.com/nspcc-dev/dbft/internal/crypto"
+
```
Comment (roman-khimov) [inline internal/block/block_test.go:13]: Ugh.

File: internal/crypto/ecdsa.go [hunk 11]
```diff
@@ -5,10 +5,11 @@ import (
 	"crypto/elliptic"
 	"crypto/sha256"
 	"errors"
-	"github.com/nspcc-dev/dbft"
 	"io"
 	"math/big"
 
+	"github.com/nspcc-dev/dbft"
+
```
Comment (roman-khimov) [inline internal/crypto/ecdsa.go:12]: :eyes:

**Comment by roman-khimov** [comment]
It looks OK except for some new exports. I'd then want to simplify payload interfaces (getters/setters suck), but that can wait.

File: helpers.go (hunk 9 above)
Comment (AnnaShaleva) [reply helpers.go]: > This I don't like. It's an internal thing.

I completely understand, but otherwise it's required to implement the whole Payload/Hash/Address interfaces on new structs, I thought it doesn't worth the effort. Fixed.

> Does anything change wrt package size if you leave it as is

Yes, exported version is larger:
```
-rw-rw-r-- 1 anna anna 188828 мар  4 12:44 build.exported
-rw-rw-r-- 1 anna anna 188676 мар  4 12:44 build.private
```

File: block/block_test.go [hunk 12]
```diff
@@ -6,6 +6,7 @@ import (
 	"encoding/binary"
 	"encoding/gob"
 	"errors"
+	"github.com/nspcc-dev/dbft"
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply block/block_test.go]: Hope that I managed to fix this IDE problem.

**Comment by AnnaShaleva** [comment]
> I'd then want to simplify payload interfaces (getters/setters suck), but that can wait.

We may do this in #84.

**Review by roman-khimov — APPROVED** [review]
